tools/cp-utility/src/main.rs

```diff
@@ -0,0 +1,338 @@
+fn parse_args(args: Vec<String>) -> io::Result<CopyOperation> {
+    if !((args.len() == 4 || args.len() == 5 && args[2].eq("-a")) && args[1].eq("cp")) {
+        return Err(io::Error::new(
+            io::ErrorKind::InvalidInput,
+            "Invalid parameters. Expected cp [-a] <source> <destination>",
+        ));
+    }
+
+    if args.len() == 5 {
+        return Ok(CopyOperation {
+            source: PathBuf::from(&args[3]),
+            destination: PathBuf::from(&args[4]),
+            copy_type: CopyType::Archive,
+        });
+    }
+
+    Ok(CopyOperation {
+        source: PathBuf::from(&args[2]),
+        destination: PathBuf::from(&args[3]),
+        copy_type: CopyType::SingleFile,
+    })
+}
+
+/// Execute the copy operation
+fn do_copy(operation: CopyOperation) -> io::Result<()> {
+    match operation.copy_type {
+        CopyType::Archive => copy_recursive(&operation.source, &operation.destination)?,
+        CopyType::SingleFile => fs::copy(&operation.source, &operation.destination).map(|_| ())?,
+    };
+    Ok(())
+}
+
+/// Execute the recursive type of copy operation
+fn copy_recursive(source: &Path, dest: &Path) -> io::Result<()> {
+    // This will cover the case in which the destination exists
+    let sanitized_dest: PathBuf = if dest.exists() {
+        dest.to_path_buf()
+            .join(source.file_name().ok_or(io::Error::new(
+                io::ErrorKind::InvalidInput,
+                "Invalid source
+        file",
+            ))?)
+    } else {
+        dest.to_path_buf()
+    };
+
+    let mut stack = VecDeque::new();
+    stack.push_back((source.to_path_buf(), sanitized_dest));
+
+    while let Some((current_source, current_dest)) = stack.pop_back() {
+        if current_source.is_symlink() {
+            let target = current_source.read_link()?;
+            unix::fs::symlink(target, &current_dest)?;
+        } else if current_source.is_dir() {
+            fs::create_dir(&current_dest)?;
+            for entry in fs::read_dir(current_source)? {
+                let next_source = entry?.path();
+                let next_dest =
+                    current_dest
+                        .clone()
+                        .join(next_source.file_name().ok_or(io::Error::new(
+                            io::ErrorKind::InvalidInput,
+                            "Invalid source file",
+                        ))?);
+                stack.push_back((next_source, next_dest));
+            }
+        } else if current_source.is_file() {
+            fs::copy(current_source, current_dest)?;
+        }
+    }
+    Ok(())
+}
+
+fn main() {
+    let args: Vec<String> = env::args().collect();
+
+    let operation = parse_args(args).unwrap_or_else(|err| {
+        eprintln!("Error parsing arguments: {err}");
+        process::exit(1);
+    });
+
+    do_copy(operation).unwrap_or_else(|err| {
+        eprintln!("Error copying files: {err}");
+        process::exit(1);
```

Comment:
```suggestion
        process::exit(2);
```

Distinct error codes for distinct errors.